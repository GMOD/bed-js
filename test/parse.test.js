import BED from '../src'

describe('BED parser', () => {
  let p
  beforeAll(() => {
    p = new BED()
  })

  it('BED3', () => {
    const f1 = p.parseLine('contigA\t10875\t10884')
    const f2 = p.parseLine('co%2CtigA\t10875\t10884')
    expect(f1).toMatchSnapshot()
    expect(f2).toMatchSnapshot()
  })

  it('BED6', () => {
    const f1 = p.parseLine('contigA\t10875\t10884\ttest\t0.50\t-')
    const f2 = p.parseLine('co%2CtigA\t10875\t10884\ttest2\t0\t+')
    const f3 = p.parseLine('cotigA\t10875\t10884\ttest2\t.\t.')
    expect(f1).toMatchSnapshot()
    expect(f2).toMatchSnapshot()
    expect(f3).toMatchSnapshot()
    expect(f2.chrom).toEqual('co,tigA')
    expect(f2.chromStart).toEqual(10875)
    expect(f2.chromEnd).toEqual(10884)
  })
  it('BED12', () => {
    const f1 = p.parseLine(
      'chr22\t1000\t5000\tcloneA\t960\t+\t1000\t5000\t0\t2\t567,488,\t0,3512',
    )
    const f2 = p.parseLine(
      'chr22\t2000\t6000\tcloneB\t900\t-\t2000\t6000\t0\t2\t433,399,\t0,3601',
    )
    expect(f1).toMatchSnapshot()
    expect(f2).toMatchSnapshot()
  })
})

describe('bigGenePred', () => {
  let p
  beforeAll(() => {
    p = new BED({ type: 'bigGenePred' })
  })
  it('bigGenePred', () => {
    const f1 = p.parseLine(
      'chr1\t11868\t14409\tENST00000456328.2\t1000\t+\t11868\t11868\t255,128,0\t3\t359,109,1189,\t0,744,1352,\tDDX11L1\tnone\tnone\t-1,-1,-1,\tnone\tENST00000456328.2\tDDX11L1\tnone',
    )
    const f2 = p.parseLine(
      'chr1\t14403\t29570\tENST00000488147.1\t1000\t-\t14403\t14403\t255,128,0\t11\t98,34,152,159,198,136,137,147,99,154,37,\t0,601,1392,2203,2454,2829,3202,3511,3864,10334,15130,\tWASH7P\tnone\tnone\t-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,\tnone\tENST00000488147.1\tWASH7P\tnone',
    )
    expect(f1).toMatchSnapshot()
    expect(f2).toMatchSnapshot()
  })
})

describe('bigPsl', () => {
  let p
  beforeAll(() => {
    p = new BED({ type: 'bigPsl' })
  })
  it('bigPsl', () => {
    const f1 = p.parseLine(
      'chr1\t11873\t14361\tmAM992877\t1000\t+\t11873\t14361\t0\t3\t354,109,1141,\t0,739,1347,\t0\t1604\t+\t1604\t0,354,463,\t\t\t248956422\t1579\t25\t0\t0\t0',
    )
    const f2 = p.parseLine(
      'chr1\t11873\t14361\tmAM992877\t1000\t+\t12622\t13259\t0\t3\t354,109,1141,\t0,739,1347,\t0\t1604\t+\t1604\t0,354,463,\tcttgccgtcagccttttctttgacctcttctttctgttcatgtgtatctgctgtctcttagcccagacttcccgtgtcctttccaccgggcctttgggaggtcacagggtcttgatgctgtggtcttgatctgcaggtgtctgacttccagcaactgctggcctgtgccagggtggaagctgagcactggagtggagttttcctgtggagaggagccatgcctagagtgggatgggccattgttcatattctggcccctgttgtctgcatgtaacctaataccacgaccaggcatgggggaaagattggaggaaagttgagtgagaggatcaacttctctgacaacctaggccagtgtgtggtgatgccaggcatgcccttccccagcatcaggtctccagagctgcagaagacgacggccgacttggatcacactcttgtgagtgtccccagtgttgcagaggcagggccatcaggcaccaaagggattctgccagcatagtgctcctggaccagtgatacacccggcaccctgtcctggacaggctgttggcctggatctgagccctcgtggaggtcaaagccacctttggttctgccattactgctgtgtggaagttcactcctgccttttcctttcccgagagcctccaccaccccgagatcgcatttctcactgccttttgtctgcccagtttcaccagaagtaggcctcttcctgacaggcagctgcaccactgcctggcgctgcgcccttcctttgctctgcccgctggagacggtgtttgtcatgggcctggtctgcagggatcctgctacaaaggtgaaacccaggagagtgtggagtccagagtgttgccaggacccaggcacaggcattagtgcccgttggagaaaacaggggaatcccgaagaaatggtgggttttggccatccgtgagatcttcccagggcagctcccctctgtggaatccaatctgtcttccatcctgcgtggccgagggccaggcttctcactgggcctctgcaggaggctgccatttgtcctgcccaccgtcttagaagcgagacggagcagactcatctgctactgccctttctataataactaaagttagctgccctggactattcaccccctagtctcaatttaaaaagatccccatggccacagggcccctgcctgggggcttgtcacctcccccaccttcttcctgagtcacttctgcagccttgctccctaacctgccccacagccttgcctggatttctatctccctggcttggtgccagttcctccaagtcgatggcacctccctccctctcaaccacttgagcaaactccaagacatcttctaccccaacaccagcaattgtgccaagggccattaggctctcagcatgactatttttagagaccccgtgtctgtcactgaaaccttttttgtgggagactattcctcccatctgcaacagctgcccctgctgacggcccttctctcctccctctcatcccagagaaacaggtcagctgggagcttctgcccccactgcctagggaccaacaggggcaggaggcagtcactgaccccgagacgtttgca\t365..502\t248956422\t1579\t25\t0\t0\t1',
    )
    expect(f1).toMatchSnapshot()
    expect(f2).toMatchSnapshot()
  })
})
// bigMaf
//
describe('bigMaf', () => {
  let p
  beforeAll(() => {
    p = new BED({ type: 'bigMaf' })
  })
  it('bigMaf', () => {
    const f1 = p.parseLine(
      'chr22_KI270731v1_random\t0\t2\ta score=60.000000;s hg38.chr22_KI270731v1_random        0 2 +   150754 AC;s canFam3.chr26                27974205 2 + 38964690 AT;',
    )
    expect(f1).toMatchSnapshot()
  })
})

describe('mafSummary', () => {
  let p
  beforeAll(() => {
    p = new BED({ type: 'mafSummary' })
  })
  it('mafSummary', () => {
    const f1 = p.parseLine(
      'chr22_KI270731v1_random\t0\t9270\tcanFam3\t0.568679\t\t',
    )
    expect(f1).toMatchSnapshot()
  })
})

describe('mafFrames', () => {
  let p
  beforeAll(() => {
    p = new BED({ type: 'mafFrames' })
  })

  it('mafFrames', () => {
    const f1 = p.parseLine(
      'chr22_KI270731v1_random\t11287\t12574\thg38\t0\t-\tENST00000619792.1\t-1\t-1\t1\t1',
    )
    expect(f1).toMatchSnapshot()
  })
})

describe('bigInteract', () => {
  let p
  beforeAll(() => {
    p = new BED({ type: 'bigInteract' })
  })

  it('bigInteract', () => {
    const f1 = p.parseLine(
      'chr3\t63702628\t63705638\t.\t584\t10\t.\t0\tchr17\t58878552\t58880897\t.\t.\tchr3\t63702628\t63705638\t.\t.',
    )
    expect(f1).toMatchSnapshot()
  })
})

describe('bigChain', () => {
  let p
  beforeAll(() => {
    p = new BED({ type: 'bigChain' })
  })

  it('bigChain', () => {
    const f1 = p.parseLine(
      'chr22_KI270731v1_random\t133689\t133710\t530549\t1000\t-\t150754\tchr16\t98207768\t80144720\t80144725\t66',
    )
    expect(f1).toMatchSnapshot()
  })
})

describe('bigLink', () => {
  let p
  beforeAll(() => {
    p = new BED({ type: 'bigLink' })
  })

  it('bigLink', () => {
    const f1 = p.parseLine(
      'chr22_KI270731v1_random\t7917\t7918\t2912\t75185566',
    )
    expect(f1).toMatchSnapshot()
  })
})

describe('bigNarrowPeak', () => {
  let p
  beforeAll(() => {
    p = new BED({ type: 'bigNarrowPeak' })
  })

  it('bigNarrowPeak', () => {
    const f1 = p.parseLine(
      'chr1\t566753\t566953\t.\t468\t.\t103.84\t5.54347\t4.80079\t154',
    )
    expect(f1).toMatchSnapshot()
  })
})
describe('bigNarrowPeak with regularize', () => {
  let p
  beforeAll(() => {
    p = new BED({ type: 'bigNarrowPeak' })
  })

  it('bigNarrowPeak', () => {
    const f1 = p.parseLine(
      'chr1\t566753\t566953\t.\t468\t.\t103.84\t5.54347\t4.80079\t154',
      { regularize: true },
    )
    expect(f1).toMatchSnapshot()
  })
})
test('real world', () => {
  const autoSql = `table hg18KGchr7
    "UCSC Genes for chr7 with color plus GeneSymbol and SwissProtID"
    (
    string  chrom;              "Reference sequence chromosome or scaffold"
    uint    chromStart; "Start position of feature on chromosome"
    uint    chromEnd;   "End position of feature on chromosome"
    string  name;               "Name of gene"
    uint    score;              "Score"
    char[1] strand;             "+ or - for strand"
    uint    thickStart; "Coding region start"
    uint    thickEnd;   "Coding region end"
    uint        reserved;       "Green on + strand, Red on - strand"
    string  geneSymbol; "Gene Symbol"
    string  spID;               "SWISS-PROT protein Accession number"
    )`
  const p = new BED({ autoSql })
  expect(p.autoSql).toMatchSnapshot()
})

test('generate README output', () => {
  const p = new BED()
  const f1 = p.parseLine('chr1\t0\t100')
  // outputs { chrom: 'chr1', chromStart: 0, chromEnd: 100 }
  const f2 = p.parseLine('chr1\t0\t100', { regularize: true, uniqueId: 1 })
  // outputs { uniqueId: 1, refName: 'chr1', start: 0, end: 100 }
  console.log(f1)
  console.log(f2)
  // console.log(p.parseLine('chr1\t0\t100'))
  // const p2 = new BED({ type: 'bigGenePred' })
  // const line =
  //   'chr1\t11868\t14409\tENST00000456328.2\t1000\t+\t11868\t11868\t255,128,0\t3\t359,109,1189,\t0,744,1352,\tDDX11L1\tnone\tnone\t-1,-1,-1,\tnone\tENST00000456328.2\tDDX11L1\tnone'
  // console.log(p2.parseLine(line))
})
