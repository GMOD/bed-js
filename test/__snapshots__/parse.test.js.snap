// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BED parser BED3 1`] = `
Object {
  "end": 10884,
  "refID": "contigA",
  "start": 10875,
  "strand": 0,
}
`;

exports[`BED parser BED3 2`] = `
Object {
  "end": 10884,
  "refID": "co,tigA",
  "start": 10875,
  "strand": 0,
}
`;

exports[`BED parser BED6 1`] = `
Object {
  "end": 10884,
  "name": "test",
  "refID": "contigA",
  "score": 0.5,
  "start": 10875,
  "strand": -1,
}
`;

exports[`BED parser BED6 2`] = `
Object {
  "end": 10884,
  "name": "test2",
  "refID": "co,tigA",
  "score": 0,
  "start": 10875,
  "strand": 1,
}
`;

exports[`BED parser BED6 3`] = `
Object {
  "end": 10884,
  "name": "test2",
  "refID": "cotigA",
  "start": 10875,
  "strand": 0,
}
`;

exports[`bigChain bigChain 1`] = `
Object {
  "chain_score": 66,
  "end": 133710,
  "name": "530549",
  "q_end": 80144725,
  "q_name": "chr16",
  "q_size": 98207768,
  "q_start": 80144720,
  "refID": "chr22_KI270731v1_random",
  "score": 1000,
  "start": 133689,
  "strand": -1,
  "t_size": 150754,
}
`;

exports[`bigGenePred bigGenePred 1`] = `
Object {
  "block_count": 3,
  "block_sizes": Array [
    359,
    109,
    1189,
  ],
  "cds_end_stat": "none",
  "cds_start_stat": "none",
  "chrom_starts": Array [
    0,
    744,
    1352,
  ],
  "end": 14409,
  "exon_frames": Array [
    -1,
    -1,
    -1,
  ],
  "gene_name": "ENST00000456328.2",
  "gene_name2": "DDX11L1",
  "gene_type": "none",
  "name": "ENST00000456328.2",
  "name2": "DDX11L1",
  "refID": "chr1",
  "reserved": "255,128,0",
  "score": 1000,
  "start": 11868,
  "strand": 1,
  "thick_end": 11868,
  "thick_start": 11868,
  "type": "none",
}
`;

exports[`bigGenePred bigGenePred 2`] = `
Object {
  "block_count": 11,
  "block_sizes": Array [
    98,
    34,
    152,
    159,
    198,
    136,
    137,
    147,
    99,
    154,
    37,
  ],
  "cds_end_stat": "none",
  "cds_start_stat": "none",
  "chrom_starts": Array [
    0,
    601,
    1392,
    2203,
    2454,
    2829,
    3202,
    3511,
    3864,
    10334,
    15130,
  ],
  "end": 29570,
  "exon_frames": Array [
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
    -1,
  ],
  "gene_name": "ENST00000488147.1",
  "gene_name2": "WASH7P",
  "gene_type": "none",
  "name": "ENST00000488147.1",
  "name2": "WASH7P",
  "refID": "chr1",
  "reserved": "255,128,0",
  "score": 1000,
  "start": 14403,
  "strand": -1,
  "thick_end": 14403,
  "thick_start": 14403,
  "type": "none",
}
`;

exports[`bigInteract bigInteract 1`] = `
Object {
  "color": "0",
  "end": 63705638,
  "refID": "chr3",
  "score": 584,
  "source_chrom": "chr17",
  "source_end": 58880897,
  "source_start": 58878552,
  "start": 63702628,
  "strand": 0,
  "target_chrom": "chr3",
  "target_end": 63705638,
  "target_start": 63702628,
  "value": "10",
}
`;

exports[`bigLink bigLink 1`] = `
Object {
  "end": 7918,
  "name": "2912",
  "q_start": 75185566,
  "refID": "chr22_KI270731v1_random",
  "start": 7917,
  "strand": 0,
}
`;

exports[`bigMaf bigMaf 1`] = `
Object {
  "end": 2,
  "maf_block": "a score=60.000000;s hg38.chr22_KI270731v1_random        0 2 +   150754 AC;s canFam3.chr26                27974205 2 + 38964690 AT;",
  "refID": "chr22_KI270731v1_random",
  "start": 0,
  "strand": 0,
}
`;

exports[`bigNarrowPeak bigNarrowPeak 1`] = `
Object {
  "end": 566953,
  "p_value": 5.54347,
  "peak": 154,
  "q_value": 4.80079,
  "refID": "chr1",
  "score": 468,
  "signal_value": 103.84,
  "start": 566753,
  "strand": 0,
}
`;

exports[`bigPsl bigPsl 1`] = `
Object {
  "block_count": 3,
  "block_sizes": Array [
    354,
    109,
    1141,
  ],
  "chrom_size": 248956422,
  "chrom_starts": Array [
    0,
    739,
    1347,
  ],
  "end": 14361,
  "match": 1579,
  "mis_match": 25,
  "n_count": 0,
  "name": "mAM992877",
  "o_cds": "",
  "o_chrom_end": 1604,
  "o_chrom_size": 1604,
  "o_chrom_start": 0,
  "o_chrom_starts": Array [
    0,
    354,
    463,
  ],
  "o_sequence": "",
  "o_strand": Array [
    "+",
  ],
  "refID": "chr1",
  "rep_match": 0,
  "reserved": 0,
  "score": 1000,
  "seq_type": 0,
  "start": 11873,
  "strand": 1,
  "thick_end": 14361,
  "thick_start": 11873,
}
`;

exports[`bigPsl bigPsl 2`] = `
Object {
  "block_count": 3,
  "block_sizes": Array [
    354,
    109,
    1141,
  ],
  "chrom_size": 248956422,
  "chrom_starts": Array [
    0,
    739,
    1347,
  ],
  "end": 14361,
  "match": 1579,
  "mis_match": 25,
  "n_count": 0,
  "name": "mAM992877",
  "o_cds": "365..502",
  "o_chrom_end": 1604,
  "o_chrom_size": 1604,
  "o_chrom_start": 0,
  "o_chrom_starts": Array [
    0,
    354,
    463,
  ],
  "o_sequence": "cttgccgtcagccttttctttgacctcttctttctgttcatgtgtatctgctgtctcttagcccagacttcccgtgtcctttccaccgggcctttgggaggtcacagggtcttgatgctgtggtcttgatctgcaggtgtctgacttccagcaactgctggcctgtgccagggtggaagctgagcactggagtggagttttcctgtggagaggagccatgcctagagtgggatgggccattgttcatattctggcccctgttgtctgcatgtaacctaataccacgaccaggcatgggggaaagattggaggaaagttgagtgagaggatcaacttctctgacaacctaggccagtgtgtggtgatgccaggcatgcccttccccagcatcaggtctccagagctgcagaagacgacggccgacttggatcacactcttgtgagtgtccccagtgttgcagaggcagggccatcaggcaccaaagggattctgccagcatagtgctcctggaccagtgatacacccggcaccctgtcctggacaggctgttggcctggatctgagccctcgtggaggtcaaagccacctttggttctgccattactgctgtgtggaagttcactcctgccttttcctttcccgagagcctccaccaccccgagatcgcatttctcactgccttttgtctgcccagtttcaccagaagtaggcctcttcctgacaggcagctgcaccactgcctggcgctgcgcccttcctttgctctgcccgctggagacggtgtttgtcatgggcctggtctgcagggatcctgctacaaaggtgaaacccaggagagtgtggagtccagagtgttgccaggacccaggcacaggcattagtgcccgttggagaaaacaggggaatcccgaagaaatggtgggttttggccatccgtgagatcttcccagggcagctcccctctgtggaatccaatctgtcttccatcctgcgtggccgagggccaggcttctcactgggcctctgcaggaggctgccatttgtcctgcccaccgtcttagaagcgagacggagcagactcatctgctactgccctttctataataactaaagttagctgccctggactattcaccccctagtctcaatttaaaaagatccccatggccacagggcccctgcctgggggcttgtcacctcccccaccttcttcctgagtcacttctgcagccttgctccctaacctgccccacagccttgcctggatttctatctccctggcttggtgccagttcctccaagtcgatggcacctccctccctctcaaccacttgagcaaactccaagacatcttctaccccaacaccagcaattgtgccaagggccattaggctctcagcatgactatttttagagaccccgtgtctgtcactgaaaccttttttgtgggagactattcctcccatctgcaacagctgcccctgctgacggcccttctctcctccctctcatcccagagaaacaggtcagctgggagcttctgcccccactgcctagggaccaacaggggcaggaggcagtcactgaccccgagacgtttgca",
  "o_strand": Array [
    "+",
  ],
  "refID": "chr1",
  "rep_match": 0,
  "reserved": 0,
  "score": 1000,
  "seq_type": 1,
  "start": 11873,
  "strand": 1,
  "thick_end": 13259,
  "thick_start": 12622,
}
`;

exports[`mafFrames mafFrames 1`] = `
Object {
  "end": 12574,
  "frame": "0",
  "is_exon_end": "1",
  "is_exon_start": "1",
  "name": "ENST00000619792.1",
  "next_frame_pos": -1,
  "prev_frame_pos": -1,
  "refID": "chr22_KI270731v1_random",
  "src": "hg38",
  "start": 11287,
  "strand": -1,
}
`;

exports[`mafSummary mafSummary 1`] = `
Object {
  "end": 9270,
  "left_status": Array [],
  "refID": "chr22_KI270731v1_random",
  "right_status": Array [],
  "score": 0.568679,
  "src": "canFam3",
  "start": 0,
  "strand": 0,
}
`;
